test/e2e/vmagent_test.go

```diff
@@ -291,3 +302,411 @@ func createExternalNodeCRD(data *TestData, nodeName string, ifName string, ip st
 	testEn.AddInterface(ifName, ipList)
 	return data.crdClient.CrdV1alpha1().ExternalNodes(namespace).Create(context.TODO(), testEn.Get(), metav1.CreateOptions{})
 }
+
+func testExternalNodeWithANP(t *testing.T, data *TestData, vmList []vmInfo) {
+	if len(vmList) < 2 {
+		t.Skipf("Skipping test as it requires 2 different VMs but the setup has %d", len(vmList))
+	}
+	// Add labels on the VMs.
+	addLabelsOnExternalNode(t, data, namespace, vmList)
+	t.Run("testANPOnLinuxVM", func(t *testing.T) { testANPOnVMs(t, data, vmList, linuxOS) })
+	t.Run("testANPOnWindowsVM", func(t *testing.T) { testANPOnVMs(t, data, vmList, windowsOS) })
+}
+
+func testANPOnVMs(t *testing.T, data *TestData, vmList []vmInfo, osType string) {
+	appliedToVM, peerVM, err := getVMsByOSType(vmList, osType)
+	if err != nil {
+		t.Skipf("Skip case testANPOnVMs: %v", err)
+	}
+	// Test TCP rules in ANP
+	t.Run("testANPOnExternalNodeWithTCP", func(t *testing.T) {
+		// Use ExternalEntity in an ingress rule configuration.
+		assert.Nil(t, testANPWithConnectivity(t, data, "anp-vmagent-ingress-tcp-entity", namespace, appliedToVM, peerVM, ProtocolTCP, true, crdv1alpha1.RuleActionDrop, true))
+		// Use IP in an egress rule configuration.
+		assert.Nil(t, testANPWithConnectivity(t, data, "anp-vmagent-egress-tcp-ip", namespace, appliedToVM, peerVM, ProtocolTCP, false, crdv1alpha1.RuleActionDrop, false))
+	})
+	// Test UDP rules in ANP
+	t.Run("testANPOnExternalNodeWithUDP", func(t *testing.T) {
+		assert.Nil(t, testANPWithConnectivity(t, data, "anp-vmagent-ingress-udp-entity", namespace, appliedToVM, peerVM, ProtocolUDP, true, crdv1alpha1.RuleActionReject, false))
+	})
+	// Test ICMP rules in ANP
+	t.Run("testANPOnExternalNodeWithICMP", func(t *testing.T) {
+		assert.Nil(t, testANPWithConnectivity(t, data, "anp-vmagent-ingress-icmp-ip", namespace, appliedToVM, peerVM, ProtocolICMP, true, crdv1alpha1.RuleActionDrop, false))
+	})
+	// Test FQDN rules in ANP
+	t.Run("testANPOnExternalNodeWithFQDN", func(t *testing.T) {
+		assert.Nil(t, testANPWithFQDN(t, data, "anp-vmagent-fqdn", namespace, appliedToVM, []string{"www.facebook.com"}, []string{"docs.google.com"}, []string{"maps.google.com"}))
```

Comment:
The subtests already check error and fail themselves. Why making them return error and check again on caller side? I see `testANPWithFQDN` never returns error while `testANPWithConnectivity` returns error only when the protocol is not supported, but this is test code, input is under control so it would never happen, right?